### PR TITLE
feat: add gradual padding for gradual trends

### LINKED ISSUE #112: [Feature] Add method param gradual_padding
Same functionality as the padding option for abrupt, but now also for gradual. This just allows people interested in ITSA for looking for periods that fiat have slope changes, but they want to see sustained effects after the slope uplift as a part of their same treatment period.

Personally, at work, I've come across a few scenarios where this setting would be beneficial, e.g. short gradual change.

Of course, if anything breaks the direction of the positivity, the trend padding will end, as it currently does with abrupt behaviour.

Can start this after previous related issue #111 is sorted.

Documentation could be nice... but not fully essential. Maybe can squish in a quick cell at the end of the abrupt tutorial notebook, showing gradual + abrupt case with no padding, and also both with full padding.

### PR DESCRIPTION
**Pull request by barshadeb**
## **Summary**

Adds a `gradual_padding` method parameter to `detect_trends()`, enabling gradual Up/Down trend segments to extend into adjacent flat periods. This mirrors the existing `abrupt_padding` behaviour for gradual trends, allowing users to model sustained effects after a slope change as part of the same treatment period.
Closes #112

## **Motivation**

In ITSA and observational causal inference, a gradual slope change may represent the start of a treatment effect, but the actual sustained impact continues into the following flat period. Without padding, these flat periods are classified separately, fragmenting what is logically a single treatment window. Gradual padding lets the trend segment absorb adjacent flat periods so the full extent of the treatment is captured.

#### New behaviour
- New `gradual_padding` parameter in `detect_trends()` (default `0`, no change to existing behaviour)
- Gradual padding extends the end date of gradual Up/Down segments forward by the specified number of days
- Extension is bounded: stops before overlapping the next non-Flat segment and clamps to the dataset end date
- Padded segments receive a `padded: True` flag to distinguish them from unpadded segments
- Padded segments are protected from artifact cleanup (short/flat reclassification) so they retain their trend classification

## **Files changed**

| File Path | Description of Changes |
| :--- | :--- |
| `pytrendy/detect_trends.py` | Document and forward `gradual_padding` in `method_params` |
| `pytrendy/post_processing/segments_refine/__init__.py` | Import and invoke `pad_gradual_trends` after expand/contract, before `abrupt_shaving` |
| `pytrendy/post_processing/segments_refine/gradual_expand_contract.py` | New `pad_gradual_trends()` function with overlap avoidance, dataset-end clamping, and padded flag |
| `pytrendy/post_processing/segments_refine/artifact_cleanup.py` | Respect `padded` flag to prevent reclassification of padded segments |
| `tests/test_gradual_padding.py` | New test file with 4 test cases |

## **Tests added**

| Test Name | Verification Target |
| :--- | :--- |
| `test_gradual_padding_28` | 28-day padding extends gradual segments into flat regions correctly |
| `test_gradual_padding_168_clamps_to_nonflat` | Large padding (168 days) is truncated at the next non-Flat segment boundary |

<details>
<summary><b>📈 View Segment Transition Summary (Before vs. After)</b></summary>

#### **Before (no padding) — 9 segments:**

- `Up` (2025-01-02 → 2025-01-24)
- `Down` (2025-01-25 → 2025-02-05)
- `Flat` (2025-02-06 → 2025-02-09)
- `Up` (2025-02-10 → 2025-03-14)
- `Flat` (2025-03-15 → 2025-03-17)
- `Down` (2025-03-18 → 2025-04-01)
- `Up` (2025-04-02 → 2025-05-08)
- `Down` (2025-05-09 → 2025-06-17)
- `Flat` (2025-06-18 → 2025-06-30)

#### **After (gradual_padding=28) — 6 segments, 0 flats:**

- `Up` (2025-01-02 → 2025-01-24) — no room to pad
- `Down` (2025-01-25 → 2025-02-09) — padded into flat ✓
- `Up` (2025-02-10 → 2025-03-17) — padded into flat ✓
- `Down` (2025-03-18 → 2025-04-01) — no room to pad
- `Up` (2025-04-02 → 2025-05-08) — no room to pad
- `Down` (2025-05-09 → 2025-06-30) — padded to dataset end ✓

**The key difference:** The 3 `Flat` gaps (Feb 6–9, Mar 15–17, and Jun 18–30) are successfully absorbed into adjacent trend segments, resulting in 6 continuous trend periods instead of 9 fragmented ones.

</details>

chart_before_no_padding

chart_after_padding_28

### CODE REVIEW

File: pytrendy/post_processing/segments_refine/__init__.py [hunk 1]
```diff
@@ -41,6 +41,7 @@ def refine_segments(df: pd.DataFrame, value_col: str, segments: list[dict], meth
     segments_refined = group_segments(segments_refined) # grouping 1st pass: sporadic flats & noises
 
     segments_refined = expand_contract_segments(df, value_col, segments_refined) # for gradual
+    segments_refined = pad_gradual_trends(df, value_col, segments_refined, method_params) # for gradual padding
```
Comment (RussellSB) [inline pytrendy/post_processing/segments_refine/__init__.py:44]: I would say not to mud up this high level layer in segments refine.
Follow more of a parallel structure to abrupt - the padded logic is called from within shave_abrupt_trends for example. So follow the same, make the gradual padded logic called from within expand_contract_segments call at the end.

File: pytrendy/post_processing/segments_refine/__init__.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 import pandas as pd
 from copy import deepcopy
 from .trend_classify import classify_trends
-from .gradual_expand_contract import expand_contract_segments
+from .gradual_expand_contract import expand_contract_segments, pad_gradual_trends
```
Comment (RussellSB) [inline pytrendy/post_processing/segments_refine/__init__.py:9]: no need to import once lower comment addressed

File: tests/test_gradual_padding.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            {'direction': 'Down', 'start': '2025-01-25', 'end': '2025-02-09'},
+            {'direction': 'Up',   'start': '2025-02-10', 'end': '2025-03-17'},
+            {'direction': 'Down', 'start': '2025-03-18', 'end': '2025-04-01'},
+            {'direction': 'Up',   'start': '2025-04-02', 'end': '2025-05-08'},
+            {'direction': 'Down', 'start': '2025-05-09', 'end': '2025-06-30'},
+        ]
+        assert_segments_match(results.segments, expected_segments)
+
+    @pytest.mark.core
+    def test_gradual_padding_zero_no_change(self):
```
Comment (RussellSB) [inline tests/test_gradual_padding.py:86]: remove this test - it already exists by default as gradual core

File: tests/test_gradual_padding.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            {'direction': 'Down', 'start': '2025-01-25', 'end': '2025-02-09'},
+            {'direction': 'Up',   'start': '2025-02-10', 'end': '2025-03-17'},
+            {'direction': 'Down', 'start': '2025-03-18', 'end': '2025-04-01'},
+            {'direction': 'Up',   'start': '2025-04-02', 'end': '2025-05-08'},
+            {'direction': 'Down', 'start': '2025-05-09', 'end': '2025-06-30'},
+        ]
+        assert_segments_match(results.segments, expected_segments)
+
+    @pytest.mark.core
+    def test_gradual_padding_28_padded_flags(self):
```
Comment (RussellSB) [inline tests/test_gradual_padding.py:38]: Remove this test - not entirely important behaviour to focus on to be honest

File: tests/test_gradual_padding.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        assert results.segments[2]['padded'] is True
+        # Down (2025-03-18 to 2025-04-01): next segment starts immediately, no room to pad
+        assert results.segments[3].get('padded', False) is False
+        # Up (2025-04-02 to 2025-05-08): next segment starts immediately, no room to pad
+        assert results.segments[4].get('padded', False) is False
+        # Down (2025-05-09 to 2025-06-30): padded to dataset end
+        assert results.segments[5]['padded'] is True
+
+    @pytest.mark.core
+    def test_gradual_padding_168_clamps_to_nonflat(self):
```
Comment (RussellSB) [inline tests/test_gradual_padding.py:63]: Find another example from synthetic or test data to try out this idea on. And you can use segments in a haystack.

The idea is good, it just seems a bit redundant and unvariede to focus on the same input/output as the example above. This scenario in specific doesnt really seem to have enough leeway really for having scenarios where gradual pads all the way up to 168.

Try refer to the example referenced from the issue, one of the recent zero baseline examples for example.

File: pytrendy/post_processing/segments_refine/artifact_cleanup.py [hunk 6]
```diff
@@ -316,7 +316,7 @@ def has_partial_overlap_prev(segment: dict, segment_prev: dict) -> bool:
             segment['direction'] = 'Noise' 
             if 'trend_class' in segment: del segment['trend_class']
 
-        if trend_ends_too_close or trend_too_small or trend_too_flat:
+        if trend_ends_too_close or trend_too_small or (trend_too_flat and not is_padded):
```
Comment (RussellSB) [inline pytrendy/post_processing/segments_refine/artifact_cleanup.py:319]: why is this edit needed?

**Review by RussellSB — CHANGES_REQUESTED** [review]
Hey @barshadeb good job. Left some nitpicky suggestions for maintaining current structures, and advice for how the introduced tests could be more directed.

**Comment by barshadeb** [comment]
## Summary
Adds a `gradual_padding` method parameter to `detect_trends()`, enabling gradual Up/Down trend segments to extend into adjacent flat periods. This mirrors the existing `abrupt_padding` behaviour for gradual trends, allowing users to model sustained effects after a slope change as part of the same treatment period. Closes #112.


## Changes

### New Behaviour
* New `gradual_padding` parameter in `detect_trends()` (default `0`, maintaining existing behaviour by default).
* Gradual padding extends the end date of gradual Up/Down segments forward by the specified number of days.
* Extension is bounded: stops before overlapping the next non-Flat segment and clamps to the dataset end date.
* Padded segments receive a `padded: True` flag to distinguish them from unpadded segments.
* Padded segments are protected from artifact cleanup (short/flat reclassification) so they retain their trend classification.

### Files Changed
* `pytrendy/detect_trends.py`
* `pytrendy/post_processing/segments_refine/__init__.py`
* `pytrendy/post_processing/segments_refine/gradual_expand_contract.py`
* `pytrendy/post_processing/segments_refine/artifact_cleanup.py`
* `tests/test_gradual_padding.py`
* `tests/tests_plotting/edgecases/test_plot_pytrendy_edgecases.py`

## Structural Design
The padding logic lives inside `expand_contract_segments` rather than as a separate top-level step in `segments_refine/__init__.py`. This follows the same pattern as abrupt padding, which is called from within `shave_abrupt_trends`. The high-level orchestrator stays clean and the expand/contract module owns all gradual boundary adjustments in one place.

## Tests Added
* `test_gradual_padding_28`
* `test_gradual_padding_clamps_to_nonflat`

## Review Feedback Addressed
* **Structural refactor:** `pad_gradual_trends` folded into `expand_contract_segments` instead of being a separate step in the orchestrator — follows the abrupt padding pattern.
* **Test trimming:** Removed `test_gradual_padding_zero_no_change` (redundant with existing gradual core test) and `test_gradual_padding_28_padded_flags` (low-value).
* **Varied test data:** Replaced the synthetic clamping test with `gradual_ramp_edgecases.csv` which has a distinct structure (gradual Up → flat → abrupt Down) rather than reusing the same synthetic input/output.
* **Style:** Used dictionary literals instead of `dict()` calls; added clarifying comments on `is_padded` and the `trend_too_flat` guard explaining why padded segments need protection.

File: pytrendy/post_processing/segments_refine/__init__.py (hunk 1 above)
Comment (barshadeb) [reply pytrendy/post_processing/segments_refine/__init__.py]: Agreed, I’ve updated the padding logic. It is now called from within expand_contract_segments (line 124 of gradual_expand_contract.py), not as a separate top-level step in __init__.py. The comment on line 44 of __init__.py says "# for gradual + gradual padding.

File: pytrendy/post_processing/segments_refine/__init__.py (hunk 2 above)
Comment (barshadeb) [reply pytrendy/post_processing/segments_refine/__init__.py]: Done, the import of pad_gradual_trends has been removed. The function is now enclosed within expand_contract_segments.
Line 9 only has: from .gradual_expand_contract import expand_contract_segments

File: tests/test_gradual_padding.py (hunk 3 above)
Comment (barshadeb) [reply tests/test_gradual_padding.py]: Done, I removed test_gradual_padding_zero_no_change

File: tests/test_gradual_padding.py (hunk 4 above)
Comment (barshadeb) [reply tests/test_gradual_padding.py]: Removed!

File: pytrendy/post_processing/segments_refine/artifact_cleanup.py (hunk 6 above)
Comment (barshadeb) [reply pytrendy/post_processing/segments_refine/artifact_cleanup.py:324]: The edit protects padded segments from being immediately reclassified as "Flat" by the cleanup pass.

When gradual_padding extends a gradual trend into an adjacent flat region, the extended portion naturally looks "too flat" (the max/min may not land in the last 30% anymore). Without the and not is_padded guard, trend_too_flat would be True, and the segment would be converted right back to Flat... undoing the intentional padding the user requested.

The same applies to abrupt padding. The is_padded flag (set in _pad_gradual_trends and the abrupt equivalent) marks segments where the extension was deliberate, so the cleanup logic respects that intent rather than fighting it.

File: tests/test_gradual_padding.py (hunk 5 above)
Comment (barshadeb) [reply tests/test_gradual_padding.py]: Swapped the dataset to gradual_ramp_edgecases.csv to avoid overlapping with the same input used in test_gradual_padding_28. Kept assert_segments_match rather than switching to assert_segments_in_a_haystack. Since we know the exact expected output after clamping, an exact match is the stronger assertion here. Using haystack would weaken the test by allowing extra unexpected segments to slip through.

File: pytrendy/post_processing/segments_refine/artifact_cleanup.py (hunk 6 above)
Comment (RussellSB) [reply pytrendy/post_processing/segments_refine/artifact_cleanup.py:324]: @barshadeb  noted... and I identified an edge case that triggers this.
```
df = pd.read_csv('tests/tests_crashes_edgecases/data/gradual_ramp_edgecases.csv')
results = pt.detect_trends(df, date_col='date', value_col='gradual_ramp_90d',
                           plot=True, method_params={'gradual_padding': 168})
```

Before (Without Guard):
Image

After (With Guard):

Image

File: tests/test_gradual_padding.py (hunk 5 above)
Comment (RussellSB) [reply tests/test_gradual_padding.py]: Confirmed. This was raised in PR #170 , now accounted for.

```
df = pd.read_csv('tests/tests_crashes_edgecases/data/zero_baseline_edgecases_2.csv')
value_col = 'zero_baseline_market_entry_2'
```
### Before
```
results = pt.detect_trends(df, date_col='date', value_col=value_col,
                           plot=True, method_params=dict(gradual_padding=0))
```
image

### After
```
results = pt.detect_trends(df, date_col='date', value_col=value_col,
                           plot=True, method_params=dict(gradual_padding=28))
```
image

**Review by RussellSB — APPROVED** [review]
